controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/UpdateStreamTask.java

```diff
@@ -51,34 +52,42 @@ public UpdateStreamTask(final StreamMetadataTasks streamMetadataTasks,
         String scope = request.getScope();
         String stream = request.getStream();
 
-        return streamMetadataStore.getConfigurationRecord(scope, stream, true, context, executor)
-                .thenCompose(configProperty -> {
-                    if (!configProperty.isUpdating()) {
-                        // if the state is updating but the configuration record is not updating, we should reset the state to ACTIVE.
-                        return streamMetadataStore.resetStateConditionally(scope, stream, State.UPDATING, context, executor)
-                                .thenRun(() -> {
+        return streamMetadataStore.getVersionedState(scope, stream, context, executor)
+                .thenCompose(versionedState -> streamMetadataStore.getConfigurationRecord(scope, stream, context, executor)
+                        .thenCompose(versionedMetadata -> {
+                            if (!versionedMetadata.getObject().isUpdating()) {
+                                CompletableFuture<VersionedMetadata<State>> future = versionedState.getObject().equals(State.UPDATING) ?
+                                     streamMetadataStore.updateVersionedState(scope, stream, State.ACTIVE,
+                                            versionedState, context, executor) : CompletableFuture.completedFuture(null);
+
+                                return future.thenApply(x -> {
```

Comment:
I don't understand why do we throw this exception after doing a bunch of work.

Reply:
There isnt bunch of work per se. 

For each processing we first set a processing specific metadata record and then set the appropriate state. 
Once processing completes, we reset the metadata record followed by resetting the state. 

If we crash before resetting process related metadata record then in rerun we will find record 
to be reset while state would be set. 
At a logical level here is what we are doing:
```
1. get state
2. get record
3. if (record.is-not-set && state.is-set)
    reset state and return;
4. check if started else throw `StartExcecption` 
5. set-state 
6. run workflow
7. reset record
8. reset state
```

So if you notice, before `4` its only two `gets` from the store.